File: stringio.pas
//  This file is part of Adlib Tracker II (AT2).
//
//  AT2 is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  AT2 is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with AT2.  If not, see <http://www.gnu.org/licenses/>.

unit StringIO;
{$S-,Q-,R-,V-,B-,X+}
{$PACKRECORDS 1}
{$MODESWITCH CVAR}
{$L stringio.o}
interface

function asciiz_string(str: String): String; cdecl; external;

{$IFNDEF ADT2PLAY}

type
  tCHARSET = Set of Char;

var
  DEC_NUM_CHARSET: set of Char; cvar; external;
  HEX_NUM_CHARSET: set of Char; cvar; external;

function CStrLen(str: String): Byte; cdecl; external;
function CStr2Len(str: String): Byte; cdecl; external;
function C3StrLen(str: String): Byte; cdecl; external;

{$ENDIF} // NOT DEFINED(ADT2PLAY)

function byte2hex(value: Byte): String; cdecl; external;
function byte2dec(value: Byte): String; cdecl; external;
function Capitalize(str: String): String; cdecl; external;
function Upper(str: String): String; cdecl; external;
function Lower(str: String): String; cdecl; external;
{$IFNDEF ADT2PLAY}
function iCASE(str: String): String; cdecl; external;
function RotStrL(str1,str2: String; shift: Byte): String; cdecl; external;
function RotStrR(str1,str2: String; shift: Byte): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
function ExpStrL(str: String; size: Byte; chr: Char): String; cdecl; external;
function ExpStrR(str: String; size: Byte; chr: Char): String; cdecl; external;
{$IFNDEF ADT2PLAY}
function ExpC2StrL(str: String; size: Byte; chr: Char): String; cdecl; external;
function ExpC2StrR(str: String; size: Byte; chr: Char): String; cdecl; external;
function ExpC3StrL(str: String; size: Byte; chr: Char): String; cdecl; external;
function ExpC3StrR(str: String; size: Byte; chr: Char): String; cdecl; external;
function CenterStr(str: String; size: Byte): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
function DietStr(str: String; size: Byte): String; cdecl; external;
function CutStr(str: String): String; cdecl; external;
{$IFNDEF ADT2PLAY}
function CutStrL(str: String; margin: Byte): String; cdecl; external;
function CutStrR(str: String; margin: Byte): String; cdecl; external;
function FlipStr(str: String): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
function FilterStr(str: String; chr0,chr1: Char): String; cdecl; external;
{$IFNDEF ADT2PLAY}
function FilterStr1(str: String; chr0: Char): String; cdecl; external;
function FilterStr2(str: String; var chr0: tCHARSET; chr1: Char): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
function Num2str(num: Longint; base: Byte): String; cdecl; external;
function Str2num(str: String; base: Byte): Longint; cdecl; external;
{$IFNDEF ADT2PLAY}
function Bpm2str(bpm: Double): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

function truncate_string(str: String): String; cdecl; external;

{$IFNDEF ADT2PLAY}
function Upper_filename(str: String): String; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
function Lower_filename(str: String): String; cdecl; external;
{$IFNDEF ADT2PLAY}
function iCASE_filename(str: String): String; cdecl; external;

function SameName(mask,str: String): Boolean; cdecl; external;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

function PathOnly(path: String): String; cdecl; external;
function NameOnly(path: String): String; cdecl; external;
function BaseNameOnly(path: String): String; cdecl; external;
function ExtOnly(path: String): String; cdecl; external;

{$IFNDEF ADT2PLAY}
procedure StringIO_Init;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

procedure init_StringIO; cdecl; external;

implementation

{$IFNDEF ADT2PLAY}
procedure StringIO_Init;
begin
  init_StringIO;
end;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

begin
  init_StringIO;
end.

File: stringio.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#include "defines.h"
#include "platform.h"
#include <string.h>
#include "stringio.h"

// HINT: (FPC) S-: Stack checking (off)
// HINT: (FPC) Q-: Overflow checking (off)
// HINT: (FPC) R-: Range checking (off)
// HINT: (FPC) V-: Var-string checking (off)
// HINT: (FPC) B-: Complete boolean evaluation (off)
// HINT: (FPC) X+: Extended syntax (ON)
// HINT: (FPC) PACKRECORDS 1: Alignment of record elements (1)

#include "stringio/asciiz_string.c"

#if !ADT2PLAY

// UCS-2

#include "stringio/UCS2_case_conv_it.c"

uint8_t UCS2_upper_case_f[(UCS2_CP_SIZE + 7) / 8];
uint8_t UCS2_lower_case_f[(UCS2_CP_SIZE + 7) / 8];

uint16_t UCS2_upper_case_ct[UCS2_CP_SIZE];
uint16_t UCS2_lower_case_ct[UCS2_CP_SIZE];

// ASCII (intersects with `AT2' code page)

CharSet_t DEC_NUM_CHARSET;
CharSet_t HEX_NUM_CHARSET;

#if DEBUG

// Normal versions (depends on `DEBUG' macro)
#include "stringio/IsUpper.c"
#include "stringio/IsLower.c"
#include "stringio/UpCase.c"
#include "stringio/LoCase.c"

#endif // DEBUG

// `AT2' code page

#include "stringio/CP437_it.c"
#include "stringio/CP437_case_conv_it.c"

AT2_char_t UCS2_to_AT2_ct[UCS2_CP_SIZE];

uint8_t AT2_upper_case_f[(AT2_CP_SIZE + 7) / 8];
uint8_t AT2_lower_case_f[(AT2_CP_SIZE + 7) / 8];

AT2_char_t AT2_upper_case_ct[AT2_CP_SIZE];
AT2_char_t AT2_lower_case_ct[AT2_CP_SIZE];

// UTF-8

#include "stringio/UTF8_to_wchar.c"
#include "stringio/UTF8str_to_AT2.c"
#include "stringio/UTF8nstr_to_AT2.c"

// Mixed CP437 and `AT2' code page, use with caution

#include "stringio/CStrLen.c"
#include "stringio/CStr2Len.c"
#include "stringio/C3StrLen.c"

#endif // !ADT2PLAY

#include "stringio/byte2hex.c"
#include "stringio/byte2dec.c"
#include "stringio/Capitalize.c"
#include "stringio/Upper.c"
#include "stringio/Lower.c"
#if !ADT2PLAY
#include "stringio/iCASE.c"
#include "stringio/RotStrL.c"
#include "stringio/RotStrR.c"
#endif // !ADT2PLAY
#include "stringio/ExpStrL.c"
#include "stringio/ExpStrR.c"
#if !ADT2PLAY
#include "stringio/ExpCStrR.c"
#include "stringio/ExpC2StrL.c"
#include "stringio/ExpC2StrR.c"
#include "stringio/ExpC3StrL.c"
#include "stringio/ExpC3StrR.c"
#include "stringio/CenterStr.c"
#endif // !ADT2PLAY
#include "stringio/DietStr.c"
#include "stringio/CutStr.c"
#if !ADT2PLAY
#include "stringio/CutStrL.c"
#include "stringio/CutStrR.c"
#include "stringio/FlipStr.c"
#endif // !ADT2PLAY
#include "stringio/FilterStr.c"
#if !ADT2PLAY
#include "stringio/FilterStr1.c"
#include "stringio/FilterStr2.c"
#endif // !ADT2PLAY
#include "stringio/Num2str.c"
#include "stringio/Str2num.c"
#if !ADT2PLAY
#include "stringio/Bpm2str.c"
#include "stringio/number_with_commas.c"
#include "stringio/remove_spaces_right.c"

#include "stringio/Upper_filename.c"
#endif // !ADT2PLAY
#include "stringio/Lower_filename.c"
#if !ADT2PLAY
#include "stringio/iCASE_filename.c"

#include "stringio/FindPart.c"
#include "stringio/IsWild.c"
#include "stringio/SameName.c"
#endif // !ADT2PLAY

#include "stringio/DoDirSeparators.c"
#include "stringio/_FSplit.c"
#if !ADT2PLAY
#include "stringio/FSplit.c"
#endif // !ADT2PLAY
#include "stringio/PathOnly.c"
#include "stringio/NameOnly.c"
#include "stringio/BaseNameOnly.c"
#include "stringio/ExtOnly.c"

#include "stringio/init_StringIO.c"
